zetaclient/chains/evm/signer/gas.go

```diff
@@ -0,0 +1,87 @@
+package signer
+
+import (
+	"math/big"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+
+	"github.com/zeta-chain/zetacore/x/crosschain/types"
+)
+
+// Gas represents gas parameters for EVM transactions.
+//
+// This is pretty interesting because all EVM chains now support EIP-1559, but some chains do it in a specific way
+// https://eips.ethereum.org/EIPS/eip-1559
+// https://www.blocknative.com/blog/eip-1559-fees
+// https://github.com/bnb-chain/BEPs/blob/master/BEPs/BEP226.md (tl;dr: baseFee is always zero)
+type Gas struct {
+	Limit uint64
+
+	// MaxFeePerUnit absolute maximum we're willing to pay per unit of gas to get tx included in a block.
+	MaxFeePerUnit *big.Int
+
+	// PriorityFeePerUnit optional fee paid directly to validators.
+	PriorityFeePerUnit *big.Int
+}
+
+func (g Gas) validate() error {
+	if g.Limit == 0 {
+		return errors.New("gas limit is zero")
+	}
+
+	if g.MaxFeePerUnit == nil {
+		return errors.New("max fee per unit is nil")
+	}
+
+	if g.PriorityFeePerUnit == nil {
+		return errors.New("priority fee per unit is nil")
+	}
+
+	return nil
+}
+
+// isLegacy determines whether the gas is meant for LegacyTx{} (pre EIP-1559)
+// or DynamicFeeTx{} (post EIP-1559).
+//
+// Returns true if priority fee is <= 0.
+func (g Gas) isLegacy() bool {
+	return g.PriorityFeePerUnit.Sign() <= 1
+}
+
+func determineGas(cctx *types.CrossChainTx, priorityFee *big.Int, logger zerolog.Logger) (Gas, error) {
```

Comment:
> Does this method satisfies this invariant?


Yes; Based on https://eips.ethereum.org/EIPS/eip-1559 🔍 

```solidity
assert signer.balance >= transaction.gas_limit * transaction.max_fee_per_gas

# ensure that the user was willing to at least pay the base fee
assert transaction.max_fee_per_gas >= block.base_fee_per_gas

# ....

# The total must be the larger of the two
assert transaction.max_fee_per_gas >= transaction.max_priority_fee_per_gas

# priority fee is capped because the base fee is filled first
priority_fee_per_gas = min(transaction.max_priority_fee_per_gas, transaction.max_fee_per_gas - block.base_fee_per_gas)

# signer pays both the priority fee and the base fee
effective_gas_price = priority_fee_per_gas + block.base_fee_per_gas
signer.balance -= transaction.gas_limit * effective_gas_price
```

So `effective_gas_price` <= `tx.max_fee_per_gas`